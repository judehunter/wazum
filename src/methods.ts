import { NoInfer } from './utils';
import {
  DataType,
  IntegerDataType,
  NumericDataType,
  Variant,
  VariantExpr,
  VariantInstr,
  VariantStmt,
} from './variants';

export const local = {
  get: <T extends NumericDataType>(
    dataType: T,
    name: string,
  ): Variant.LocalGet<T> => ({
    __nodeType: 'localGet',
    dataType,
    name,
    returnType: dataType,
  }),
  set: <T extends NumericDataType>(
    dataType: T,
    name: string,
    value: VariantExpr<NoInfer<T>>,
  ): Variant.LocalSet => ({
    __nodeType: 'localSet',
    dataType,
    name,
    value,
    returnType: 'none',
  }),
};

export const add = <T extends NumericDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.Add<T> => ({
  __nodeType: 'add',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const sub = <T extends NumericDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.Sub<T> => ({
  __nodeType: 'sub',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const mul = <T extends NumericDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.Mul<T> => ({
  __nodeType: 'mul',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const divSigned = <T extends NumericDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.DivSigned<T> => ({
  __nodeType: 'divSigned',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const divUnsigned = <T extends NumericDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.DivUnsigned<T> => ({
  __nodeType: 'divUnsigned',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const remSigned = <T extends IntegerDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.RemSigned<T> => ({
  __nodeType: 'remSigned',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const remUnsigned = <T extends IntegerDataType>(
  dataType: T,
  left: VariantExpr<NoInfer<T>>,
  right: VariantExpr<NoInfer<T>>,
): Variant.RemUnsigned<T> => ({
  __nodeType: 'remUnsigned',
  dataType,
  left,
  right,
  returnType: dataType,
});

export const constant = <T extends NumericDataType>(
  dataType: T,
  value: number,
): Variant.Const<T> => ({
  __nodeType: 'const',
  value,
  dataType,
  returnType: dataType,
});

export const func = <T extends DataType>(
  name: string,
  signature: {
    params: [type: DataType, name: string][];
    locals: [type: DataType, name: string][];
    returnType: T;
  },
  body: VariantExpr<NoInfer<T>>,
): Variant.Func<T> => ({
  __nodeType: 'func',
  name,
  params: signature.params,
  locals: signature.locals,
  body,
  dataType: signature.returnType,
  exportName: null,
});

export const drop = (value: VariantExpr): Variant.Drop => ({
  __nodeType: 'drop',
  value,
  returnType: 'none',
});

export const block = <T extends DataType>(
  name: string | null,
  returnType: T,
  value: [
    ...VariantStmt[],
    Omit<VariantInstr, 'returnType'> & { returnType: NoInfer<T> },
  ],
): Variant.Block<T> => ({
  __nodeType: 'block',
  name,
  value,
  returnType,
});

export const call = <T extends DataType>(
  name: string,
  returnType: T,
  args: VariantExpr[],
): Variant.Call<T> => ({
  __nodeType: 'call',
  name,
  args,
  returnType,
});

export const callIndirect = <T extends DataType>(
  tableName: string,
  address: VariantExpr,
  signature: {
    params: [type: DataType, name?: string][];
    returnType: T;
  },
  args: VariantExpr[],
): Variant.CallIndirect<T> => ({
  __nodeType: 'callIndirect',
  tableName,
  address,
  params: signature.params,
  args: args,
  dataType: signature.returnType,
});

export const load = <T extends DataType>(
  
): Variant.Load<T> => ({
  __nodeType: 'load',

})