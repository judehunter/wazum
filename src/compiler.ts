import { Module } from './module';
import {
  Add,
  Block,
  Call,
  CallIndirect,
  Const,
  DataType,
  DivSigned,
  DivUnsigned,
  Drop,
  IntegerDataType,
  Load,
  LocalGet,
  LocalSet,
  LocalTee,
  Mul,
  NumericDataType,
  RemSigned,
  RemUnsigned,
  Store,
  Sub,
  Instr,
  GlobalGet,
  GlobalSet,
  GlobalTee,
  PresentDataType,
  Loop,
  BranchIf,
  Branch,
  EqualZero,
  Load8ZeroExt,
  Load8SignExt,
  Load16SignExt,
  Load16ZeroExt,
  Load32SignExt,
  Load32ZeroExt,
  Store8,
  Store16,
  Store32,
  Equal,
  GreaterEqualSigned,
  GreaterEqualUnsigned,
  LessEqualSigned,
  LessEqualUnsigned,
  GreaterThanSigned,
  GreaterThanUnsigned,
} from './nodes';

const compileSExpression = (
  sExpr: {
    fn: string;
    inlineArgs?: (string | null)[];
    blockArgs?: (string | null)[];
  },
  indent: number,
) => {
  let ret = space(indent) + `(${sExpr.fn}`;
  const filteredInlineArgs = sExpr.inlineArgs?.filter((x) => x !== null);
  const filteredBlockArgs = sExpr.blockArgs?.filter((x) => x !== null);
  if (filteredInlineArgs?.length) {
    ret += ` `;
    ret += filteredInlineArgs.join(' ');
  }
  if (filteredBlockArgs?.length) {
    ret += `\n`;
    ret += filteredBlockArgs.join('\n');
    ret += `\n` + space(indent) + `)`;
  } else {
    ret += `)`;
  }
  return ret;
};

const space = (indent: number) => ' '.repeat(indent);

const localGet = (node: LocalGet<NumericDataType>, indent = 0) => {
  return compileSExpression(
    { fn: 'local.get', inlineArgs: [`$${node.name}`] },
    indent,
  );
};

const localSet = (node: LocalSet, indent = 0) => {
  return compileSExpression(
    {
      fn: 'local.set',
      inlineArgs: [`$${node.name}`],
      blockArgs: [instr(node.value, indent + 1)],
    },
    indent,
  );
};

const localTee = (node: LocalTee<NumericDataType>, indent = 0) => {
  return compileSExpression(
    {
      fn: 'local.tee',
      inlineArgs: [`$${node.name}`],
      blockArgs: [instr(node.value, indent + 1)],
    },
    indent,
  );
};

const globalGet = (node: GlobalGet<NumericDataType>, indent = 0) => {
  return compileSExpression(
    { fn: 'global.get', inlineArgs: [`$${node.name}`] },
    indent,
  );
};

const globalSet = (node: GlobalSet, indent = 0) => {
  return compileSExpression(
    {
      fn: 'global.set',
      inlineArgs: [`$${node.name}`],
      blockArgs: [instr(node.value, indent + 1)],
    },
    indent,
  );
};

const globalTee = (node: GlobalTee<NumericDataType>, indent = 0) => {
  return compileSExpression(
    {
      fn: 'global.tee',
      inlineArgs: [`$${node.name}`],
      blockArgs: [instr(node.value, indent + 1)],
    },
    indent,
  );
};

const makeBinaryCompiler =
  <T extends { dataType: DataType; left: Instr; right: Instr }>(op: string) =>
  (node: T, indent = 0) => {
    return compileSExpression(
      {
        fn: `${node.dataType}.${op}`,
        blockArgs: [
          instr(node.left, indent + 1),
          instr(node.right, indent + 1),
        ],
      },
      indent,
    );
  };

const add = makeBinaryCompiler<Add<NumericDataType>>('add');
const sub = makeBinaryCompiler<Sub<NumericDataType>>('sub');
const mul = makeBinaryCompiler<Mul<NumericDataType>>('mul');
const divSigned = makeBinaryCompiler<DivSigned<NumericDataType>>('div_s');
const divUnsigned = makeBinaryCompiler<DivUnsigned<NumericDataType>>('div_u');
const remSigned = makeBinaryCompiler<RemSigned<NumericDataType>>('rem_s');
const remUnsigned = makeBinaryCompiler<RemUnsigned<NumericDataType>>('rem_u');
const equal = makeBinaryCompiler<Equal<NumericDataType>>('eq');
const greaterThanSigned =
  makeBinaryCompiler<GreaterThanSigned<NumericDataType>>('gt_s');
const greaterThanUnsigned =
  makeBinaryCompiler<GreaterThanUnsigned<NumericDataType>>('gt_u');
const lessThanSigned =
  makeBinaryCompiler<GreaterThanSigned<NumericDataType>>('lt_s');
const lessThanUnsigned =
  makeBinaryCompiler<GreaterThanUnsigned<NumericDataType>>('lt_u');
const greaterEqualSigned =
  makeBinaryCompiler<GreaterEqualSigned<NumericDataType>>('ge_s');
const greaterEqualUnsigned =
  makeBinaryCompiler<GreaterEqualUnsigned<NumericDataType>>('ge_u');
const lessEqualSigned =
  makeBinaryCompiler<LessEqualSigned<NumericDataType>>('le_s');
const lessEqualUnsigned =
  makeBinaryCompiler<LessEqualUnsigned<NumericDataType>>('le_u');

const constant = (node: Const<NumericDataType>, indent = 0) => {
  return compileSExpression(
    { fn: `${node.dataType}.const`, inlineArgs: [`${node.value}`] },
    indent,
  );
};

const call = (node: Call<DataType>, indent = 0) => {
  return compileSExpression(
    {
      fn: 'call',
      inlineArgs: [`$${node.name}`],
      blockArgs: node.args.map((arg) => instr(arg, indent + 1)),
    },
    indent,
  );
};

const callIndirect = (node: CallIndirect<DataType>, indent = 0) => {
  return compileSExpression(
    {
      fn: 'call_indirect',
      blockArgs: [
        space(indent + 1) + `$${node.tableName}`,
        ...node.params.map(
          (param) => space(indent + 1) + `(param ${param[0]})`,
        ),
        space(indent + 1) + `(result ${node.dataType})`,
        ...node.args.map((arg) => instr(arg, indent + 1)),
        instr(node.address, indent + 1),
      ],
    },
    indent,
  );
};

const block = (node: Block<DataType>, indent = 0) => {
  return compileSExpression(
    {
      fn: 'block',
      blockArgs: [
        node.name ? space(indent + 1) + `$${node.name}` : null,
        node.returnType !== 'none'
          ? space(indent + 1) + `(result ${node.returnType})`
          : null,
        ...node.body.map((x) => instr(x as any, indent + 1)),
      ],
    },
    indent,
  );
};

const drop = (node: Drop, indent = 0) => {
  return compileSExpression(
    { fn: 'drop', blockArgs: [instr(node.value, indent + 1)] },
    indent,
  );
};

const makeStoreCompiler =
  <T extends Store | Store8 | Store16 | Store32>(op: string) =>
  (node: T, indent = 0) => {
    return compileSExpression(
      {
        fn: `${node.dataType}.${op}`,
        blockArgs: [
          node.offset ? space(indent + 1) + `offset=${node.offset}` : null,
          node.align ? space(indent + 1) + `align=${node.align}` : null,
          instr(node.base, indent + 1),
          instr(node.value, indent + 1),
        ],
      },
      indent,
    );
  };

const store = makeStoreCompiler<Store>('store');
const store8 = makeStoreCompiler<Store8>('store8');
const store16 = makeStoreCompiler<Store16>('store16');
const store32 = makeStoreCompiler<Store32>('store32');

const makeLoadCompiler =
  <
    T extends
      | Load
      | Load8SignExt
      | Load8ZeroExt
      | Load16SignExt
      | Load16ZeroExt
      | Load32SignExt
      | Load32ZeroExt,
  >(
    op: string,
  ) =>
  (node: T, indent = 0) => {
    return compileSExpression(
      {
        fn: `${node.dataType}.${op}`,
        blockArgs: [
          node.offset ? space(indent + 1) + `offset=${node.offset}` : null,
          node.align !== null
            ? space(indent + 1) + `align=${node.align}`
            : null,
          instr(node.base, indent + 1),
        ],
      },
      indent,
    );
  };

const load = makeLoadCompiler<Load>('load');
const load8SignExt = makeLoadCompiler<Load8SignExt>('load8_s');
const load8ZeroExt = makeLoadCompiler<Load8ZeroExt>('load8_u');
const load16SignExt = makeLoadCompiler<Load16SignExt>('load16_s');
const load16ZeroExt = makeLoadCompiler<Load16ZeroExt>('load16_u');
const load32SignExt = makeLoadCompiler<Load32SignExt>('load32_s');
const load32ZeroExt = makeLoadCompiler<Load32ZeroExt>('load32_u');

const loop = (node: Loop, indent = 0) => {
  return compileSExpression(
    {
      fn: 'loop',
      inlineArgs: [node.name !== null ? `$${node.name}` : null],
      blockArgs: node.body.map((x) => instr(x, indent + 1)),
    },
    indent,
  );
};

const branchIf = (node: BranchIf, indent = 0) => {
  return compileSExpression(
    {
      fn: 'br_if',
      inlineArgs: [
        typeof node.branchTo === 'string'
          ? `$${node.branchTo}`
          : `${node.branchTo}`,
      ],
      blockArgs: [instr(node.cond, indent + 1)],
    },
    indent,
  );
};

const branch = (node: Branch, indent = 0) => {
  return compileSExpression(
    {
      fn: 'br',
      inlineArgs: [
        typeof node.branchTo === 'string'
          ? `$${node.branchTo}`
          : `${node.branchTo}`,
      ],
    },
    indent,
  );
};

const equalZero = (node: EqualZero, indent = 0) => {
  return compileSExpression(
    { fn: `${node.dataType}.eqz`, blockArgs: [instr(node.right, indent + 1)] },
    indent,
  );
};

const instrTypesCompilers = {
  localGet,
  localSet,
  localTee,
  add,
  sub,
  mul,
  divSigned,
  divUnsigned,
  remSigned,
  remUnsigned,
  constant,
  call,
  callIndirect,
  load,
  load8SignExt,
  load8ZeroExt,
  load16SignExt,
  load16ZeroExt,
  load32SignExt,
  load32ZeroExt,
  store,
  store8,
  store16,
  store32,
  globalGet,
  globalSet,
  globalTee,
  loop,
  branchIf,
  branch,
  equalZero,
  equal,
  drop,
  block,
  greaterThanSigned,
  greaterThanUnsigned,
  lessThanSigned,
  lessThanUnsigned,
  greaterEqualSigned,
  greaterEqualUnsigned,
  lessEqualSigned,
  lessEqualUnsigned,
};

const instr = (node: Instr, indent = 0): string => {
  const compiler = instrTypesCompilers[node.__nodeType];
  if (!compiler) {
    throw new Error(`Unexpected ${node.__nodeType} node`);
  }
  return compiler(node as any, indent);
};

export const compile = (m: Module) => {
  return compileSExpression(
    {
      fn: 'module',
      blockArgs: [
        ...m.memories.flatMap((mem) => [
          compileSExpression(
            {
              fn: 'memory',
              inlineArgs: [`$${mem.name}`, `${mem.initSize}`, `${mem.maxSize}`],
            },
            1,
          ),
          compileSExpression(
            {
              fn: 'export',
              inlineArgs: [`"${mem.name}"`, `(memory $${mem.name})`],
            },
            1,
          ),
          ...mem.segments.flatMap((seg) =>
            compileSExpression(
              {
                fn: 'data',
                blockArgs: [
                  instr(seg.offset, 2),
                  space(2) +
                    `"${[...seg.data]
                      .map((d) => '\\' + d.toString(16))
                      .join('')}"`,
                ],
              },
              1,
            ),
          ),
        ]),
        ...m.tables.flatMap((table) => [
          compileSExpression(
            {
              fn: 'table',
              inlineArgs: [
                `$${table.name}`,
                `${table.initSize}`,
                `${table.maxSize}`,
                `${table.type}`,
              ],
            },
            1,
          ),
          ...table.segments.flatMap((seg) =>
            compileSExpression(
              {
                fn: 'elem',
                blockArgs: [
                  // space(2) + `(table $${table.name})`,
                  instr(seg.offset, 2),
                  space(2) + seg.elems.map((x) => `$${x}`).join(' '),
                ],
              },
              1,
            ),
          ),
        ]),
        ...m.globals.map((global) =>
          compileSExpression(
            {
              fn: 'global',
              inlineArgs: [
                `$${global.name}`,
                `(${global.mutable ? 'mut ' : ''}${global.dataType})`,
              ],
              blockArgs: [instr(global.initVal, 2)],
            },
            1,
          ),
        ),
        ...m.funcs
          .filter((func) => func.exportName)
          .map((func) =>
            compileSExpression(
              {
                fn: 'export',
                inlineArgs: [`"${func.exportName}"`, `(func $${func.name})`],
              },
              1,
            ),
          ),
        ...m.funcs.map((func) =>
          compileSExpression(
            {
              fn: 'func',
              inlineArgs: [`$${func.name}`],
              blockArgs: [
                ...func.params.map(
                  ([dt, name]) => space(2) + `(param $${name} ${dt})`,
                ),
                func.dataType !== 'none'
                  ? space(2) + `(result ${func.dataType})`
                  : null,
                ...func.locals.map(
                  ([dt, name]) => space(2) + `(local $${name} ${dt})`,
                ),
                instr(func.body, 2),
              ],
            },
            1,
          ),
        ),
        m.start !== null
          ? compileSExpression(
              {
                fn: 'start',
                inlineArgs: [`$${m.start}`],
              },
              1,
            )
          : null,
      ],
    },
    0,
  );
};

export const compilers = {
  instr,
  module: compile,
  ...instrTypesCompilers,
};
