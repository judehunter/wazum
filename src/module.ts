import { compile } from './compiler';
import {
  DataType,
  Func,
  Instr,
  IntegerDataType,
  NumericDataType,
  PresentDataType,
} from './nodes';

export type Memory = {
  name: string;
  initSize: number;
  maxSize: number;
  segments: MemorySegment[];
};

export type MemorySegment = {
  data: Uint8Array;
  offset: Instr<IntegerDataType>;
};

export type Table = {
  type: 'funcref';
  name: string;
  initSize: number;
  maxSize: number;
  segments: TableSegment[];
};

export type TableSegment = {
  elems: string[];
  offset: Instr<IntegerDataType>;
};

export type Global = {
  name: string;
  dataType: DataType;
  initVal: Instr<PresentDataType>;
  mutable: boolean;
};

export class Module {
  funcs: Func<DataType>[] = [];
  memories: Memory[] = [];
  tables: Table[] = [];
  globals: Global[] = [];
  start: string | null = null;

  constructor() {}

  addFunc = (func: Func<DataType>, exported?: boolean | string) => {
    if (exported) {
      func.exportName = exported && exported === true ? func.name : exported;
    }
    this.funcs.push(func);
  };

  addMemory = (
    name: string,
    initSize: number,
    maxSize: number,
    segments: MemorySegment[],
  ) => {
    if (this.memories.length)
      throw new Error('Multiple memories not supported yet');
    this.memories.push({
      name,
      initSize,
      maxSize,
      segments,
    });
  };

  addTable = (
    name: string,
    initSize: number,
    maxSize: number,
    type: 'funcref',
    segments: TableSegment[],
  ) => {
    this.tables.push({ type, name, initSize, maxSize, segments });
  };

  addGlobal = (
    name: string,
    dataType: DataType,
    mutable: boolean,
    initVal: Instr<PresentDataType>,
  ) => {
    this.globals.push({ name, dataType, initVal, mutable });
  };

  setStart = (name: string | null) => {
    this.start = name;
  };

  compile = () => compile(this);
}
