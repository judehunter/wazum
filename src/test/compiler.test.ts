import { compilers, compile } from '../compiler';
import * as w from '../methods';
import { Module } from '../module';

describe('compiler', () => {
  const addFunc = () =>
    w.func(
      'add',
      {
        params: [
          ['i32', 'a'],
          ['i32', 'b'],
        ],
        locals: [],
        returnType: 'i32',
      },
      w.add('i32', w.local.get('i32', 'a'), w.local.get('i32', 'b')),
    );

  const addFuncImport = () =>
    w.funcImport(
      'add',
      {
        importPath: "env",
        importName: "host_add",
        params: [
          ['i32', 'a'],
          ['i32', 'b'],
        ],
        returnType: 'i32',
      }
    );

  test('clz node', () => {
    expect(
      compilers.clz(
        w.clz('i32', w.constant('i32', 1), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('ctz node', () => {
    expect(
      compilers.ctz(
        w.ctz('i32', w.constant('i32', 1), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('popcnt node', () => {
    expect(
      compilers.popcnt(
        w.popcnt('i32', w.constant('i32', 1), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('add function (local.get, i32.add, exported func)', () => {
    const m = new Module();
    m.addFunc(addFunc(), true);

    expect(compile(m)).toMatchSnapshot();
  });

  test('add function import', () => {
    const m = new Module();
    m.addFuncImport(addFuncImport());

    expect(compile(m)).toMatchSnapshot();
  });

  test('localGet node', () => {
    expect(compilers.localGet(w.local.get('f64', 'abc'))).toMatchSnapshot();
  });

  test('localSet node', () => {
    expect(
      compilers.localSet(w.local.set('f64', 'abc', w.constant('f64', 10))),
    ).toMatchSnapshot();
  });

  test('const node', () => {
    expect(compilers.constant(w.constant('f32', 1.5))).toMatchSnapshot();
  });

  test('add node', () => {
    expect(
      compilers.add(w.add('i32', w.constant('i32', 10), w.constant('i32', 5))),
    ).toMatchSnapshot();
  });

  test('sub node', () => {
    expect(
      compilers.sub(w.sub('i32', w.constant('i32', 10), w.constant('i32', 5))),
    ).toMatchSnapshot();
  });

  test('mul node', () => {
    expect(
      compilers.mul(w.mul('i32', w.constant('i32', 10), w.constant('i32', 5))),
    ).toMatchSnapshot();
  });

  test('div_s node', () => {
    expect(
      compilers.divSigned(
        w.divSigned('i32', w.constant('i32', 10), w.constant('i32', 5)),
      ),
    ).toMatchSnapshot();
  });

  test('div_u node', () => {
    expect(
      compilers.divUnsigned(
        w.divUnsigned('i32', w.constant('i32', 10), w.constant('i32', 5)),
      ),
    ).toMatchSnapshot();
  });

  test('rem_s node', () => {
    expect(
      compilers.remSigned(
        w.remSigned('i32', w.constant('i32', 10), w.constant('i32', 5)),
      ),
    ).toMatchSnapshot();
  });

  test('rem_u node', () => {
    expect(
      compilers.remUnsigned(
        w.remUnsigned('i32', w.constant('i32', 10), w.constant('i32', 5)),
      ),
    ).toMatchSnapshot();
  });

  test('and node', () => {
    expect(
      compilers.and(
        w.and('i32', w.constant('i32', 0xf0f0f0), w.constant('i32', 0x0f0f0f)),
      ),
    ).toMatchSnapshot();
  });

  test('or node', () => {
    expect(
      compilers.or(
        w.or('i32', w.constant('i32', 0xf0f0f0), w.constant('i32', 0x0f0f0f)),
      ),
    ).toMatchSnapshot();
  });

  test('xor node', () => {
    expect(
      compilers.xor(
        w.xor('i32', w.constant('i32', 0xf0f0f0), w.constant('i32', 0x0f0f0f)),
      ),
    ).toMatchSnapshot();
  });

  test('shl node', () => {
    expect(
      compilers.shl(
        w.shl('i32', w.constant('i32', 10), w.constant('i32', 5)),
      ),
    ).toMatchSnapshot();
  });

  test('shr_s node', () => {
    expect(
      compilers.shrSigned(
        w.shrSigned('i32', w.constant('i32', 20), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('shr_u node', () => {
    expect(
      compilers.shrUnsigned(
        w.shrUnsigned('i32', w.constant('i32', 20), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('rotr node', () => {
    expect(
      compilers.rotr(
        w.rotr('i32', w.constant('i32', 20), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('rotl node', () => {
    expect(
      compilers.rotl(
        w.rotl('i32', w.constant('i32', 20), w.constant('i32', 2)),
      ),
    ).toMatchSnapshot();
  });

  test('call node', () => {
    expect(
      compilers.call(w.call('abc', 'i32', [w.constant('i32', 10)])),
    ).toMatchSnapshot();
  });

  test('call_indirect node', () => {
    expect(
      compilers.callIndirect(
        w.callIndirect(
          '0',
          w.constant('i32', 0),
          { params: addFunc().params, returnType: addFunc().dataType },
          [w.constant('i32', 3), w.constant('i32', 4)],
        ),
      ),
    ).toMatchSnapshot();
  });
});
