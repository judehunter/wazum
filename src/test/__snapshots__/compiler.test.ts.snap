// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler add function (local.get, i32.add, exported func) 1`] = `
"(module
 (export "add" (func $add))
 (func $add
  (param $a i32)
  (param $b i32)
  (result i32)
  (i32.add
   (local.get $a)
   (local.get $b)
  )
 )
)"
`;

exports[`compiler add function import 1`] = `
"(module
 (import "env" "host_add" (func $add
   (param i32)
   (param i32)
   (result i32)
  )
 )
)"
`;

exports[`compiler add node 1`] = `
"(i32.add
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler and node 1`] = `
"(i32.and
 (i32.const 15790320)
 (i32.const 986895)
)"
`;

exports[`compiler call node 1`] = `
"(call $abc
 (i32.const 10)
)"
`;

exports[`compiler call_indirect node 1`] = `
"(call_indirect
 $0
 (param i32)
 (param i32)
 (result i32)
 (i32.const 3)
 (i32.const 4)
 (i32.const 0)
)"
`;

exports[`compiler clz node 1`] = `
"(i32.clz
 (i32.const 1)
 (i32.const 2)
)"
`;

exports[`compiler const node 1`] = `"(f32.const 1.5)"`;

exports[`compiler ctz node 1`] = `
"(i32.ctz
 (i32.const 1)
 (i32.const 2)
)"
`;

exports[`compiler div_s node 1`] = `
"(i32.div_s
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler div_u node 1`] = `
"(i32.div_u
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler localGet node 1`] = `"(local.get $abc)"`;

exports[`compiler localSet node 1`] = `
"(local.set $abc
 (f64.const 10)
)"
`;

exports[`compiler mul node 1`] = `
"(i32.mul
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler or node 1`] = `
"(i32.or
 (i32.const 15790320)
 (i32.const 986895)
)"
`;

exports[`compiler popcnt node 1`] = `
"(i32.popcnt
 (i32.const 1)
 (i32.const 2)
)"
`;

exports[`compiler rem_s node 1`] = `
"(i32.rem_s
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler rem_u node 1`] = `
"(i32.rem_u
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler rotl node 1`] = `
"(i32.rotl
 (i32.const 20)
 (i32.const 2)
)"
`;

exports[`compiler rotr node 1`] = `
"(i32.rotr
 (i32.const 20)
 (i32.const 2)
)"
`;

exports[`compiler shl node 1`] = `
"(i32.shl
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler shr_s node 1`] = `
"(i32.shr_s
 (i32.const 20)
 (i32.const 2)
)"
`;

exports[`compiler shr_u node 1`] = `
"(i32.shr_u
 (i32.const 20)
 (i32.const 2)
)"
`;

exports[`compiler sub node 1`] = `
"(i32.sub
 (i32.const 10)
 (i32.const 5)
)"
`;

exports[`compiler xor node 1`] = `
"(i32.xor
 (i32.const 15790320)
 (i32.const 986895)
)"
`;
